### PR TITLE
Port TypeSafeEnum and Converter to System.Text.Json

### PR DESCRIPTION
**Pull request by bartelink**
Part of #38

### CODE REVIEW

File: src/FsCodec.SystemTextJson/FsCodec.SystemTextJson.fsproj [hunk 1]
```diff
@@ -11,12 +11,14 @@
   <ItemGroup>
     <Compile Include="JsonSerializerElementExtensions.fs" />
     <Compile Include="Utf8JsonReaderExtensions.fs" />
-    <Compile Include="JsonOptionConverter.fs" />
     <Compile Include="JsonRecordConverter.fs" />
+    <Compile Include="JsonOptionConverter.fs" />
+    <Compile Include="Pickler.fs" />
```
Comment (ylibrach) [inline src/FsCodec.SystemTextJson/FsCodec.SystemTextJson.fsproj]: Should these be here?

File: src/FsCodec.SystemTextJson/TypeSafeEnumConverter.fs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+/// Utilities for working with DUs where none of the cases have a value
+module TypeSafeEnum =
+
+    let private _isTypeSafeEnum (t : Type) =
+        Union.tryGetUnion t
+        |> Option.exists (fun u -> u.cases |> Seq.forall (fun case -> case.GetFields().Length = 0))
+    let isTypeSafeEnum = memoize _isTypeSafeEnum
+
+    let tryParseT (t : Type) (str : string) =
```
Comment (ylibrach) [inline src/FsCodec.SystemTextJson/TypeSafeEnumConverter.fs]: Do you think it's worth it to memoize this function (and the others, like `toString`, etc) as well? While `Union.tryGetUnion` is memoized, these functions do call constructors, tag readers, etc each time which are relatively expensive operations - especially when taken in context of JSON documents which could be coming in at 1000s / sec.

File: src/FsCodec.SystemTextJson/FsCodec.SystemTextJson.fsproj (hunk 1 above)
Comment (bartelink) [reply src/FsCodec.SystemTextJson/FsCodec.SystemTextJson.fsproj]: hm, UnionConverter shouldnt be, will double check

File: src/FsCodec.SystemTextJson/TypeSafeEnumConverter.fs (hunk 2 above)
Comment (bartelink) [reply src/FsCodec.SystemTextJson/TypeSafeEnumConverter.fs]: I'm not averse to that happening, but would prefer to have this PR stick to the task for now
In practice, caching the lookup in the Converter might ultimately give a good solution too
Also relevant is that there is another converter (UnionConverter) which will bring some more Union.* methods into play - I'd like to do an aggressive cleanup/refactor post that point before we risk Perf-Induced-Design-Damage 😈

Comment (bartelink) [reply src/FsCodec.SystemTextJson/TypeSafeEnumConverter.fs]: (I can't imagine this not ending up being the right answer; tbh I'm also being just a touch lazy!)

File: src/FsCodec.SystemTextJson/FsCodec.SystemTextJson.fsproj (hunk 1 above)
Comment (bartelink) [reply src/FsCodec.SystemTextJson/FsCodec.SystemTextJson.fsproj]: turns out UnionConverter is where those helpers make sense to house when that converter is ported...

File: src/FsCodec.SystemTextJson/TypeSafeEnumConverter.fs (hunk 2 above)
Comment (bartelink) [reply src/FsCodec.SystemTextJson/TypeSafeEnumConverter.fs]: should probably also be applied to original; upon considering, opting to defer for now

**Pull request merged**